### ISSUE TITLE
Fallible request.dynamic

### ISSUE THREAD

**Issue by Yamakaky** [savonet/liquidsoap#784]
Is there a way to make request.dynamic fallible? Creating a request with an invalid filename seem to work, but it puts a lot of error messages in the logs.

**Comment by toots** [comment]
Yeah, that's the way for now. It's true that a nicer method would be good.

**Comment by Yamakaky** [comment]
Maybe return `""` when the queue should fail?

**Comment by smimram** [comment]
I have just added the `sleep` function which allows sleeping in the request function, so that you don't see the invalid requests too often:
```
def f() =
  sleep(10.)
  request.create("unavailable")
end

default = playlist("~/Music")

s = request.dynamic(f)
s = fallback([s,default])
out(s)
```
Is it good enough?

**Comment by Yamakaky** [comment]
Hum, I will use it for error handling when the queue is empty, so it should be good enough ^^

**Comment by toots** [comment]
I may have to review that. Is it running in the main streaming thread, blocking all streaming? Or blocking a fully thread? Duppy has facilities to sleep without blocking a thread..

**Comment by smimram** [comment]
It's `Unix.sleep` so it's blocking the current thread (but not all streaming, I tested it). But maybe there is something better to do with duppy...

**Comment by toots** [comment]
I have some ideas for this but I need to figure out if I want to add them right as we enter `beta` phase.

**Comment by Yamakaky** [comment]
We can start deciding while releasing beta. I ended up using a backup.mp3 file for error handling so I can wait.

**Comment by toots** [comment]
I want the `beta` phase to be strictly bugfix and minimal changes. The changes I have in mind here are too much for it. The truth is, the whole `playlist`/`single`/`request.{queue,equeue}`/`request.dynamic` code needs to be reviewed, there's too much in it.

**Comment by Yamakaky** [comment]
Yeah, but we can talk about the design even if we don't implement it now.
I have a setup which somewhat works using `queue` and `on_end`, but that could be improved in liquidsoap. I'll post something about, maybe on the slack.

**Comment by toots** [comment]
The way to do it right is pretty intricate. All asynchronous tasks are run via [ocaml-duppy](http://gihub.com/savonet/ocaml-duppy). The scheduler has methods to put a task back into sleep, either until it is woken up or for a defined amount of time.

The tricky part is to hook that up in the script language. I've added something similar with `server.read` recently. I just need some time to properly think it through and I would certainly take this opportunity to refactor the code of the aforementioned operators..

**Comment by toots** [comment]
Same for this one, moving it to the next major release cycle.

**Comment by smimram** [comment]
I just added an `available` parameter in #1013 which should completely solve your problem :)
